### PR TITLE
Flow: add loki.source.cloudflare component

### LINKED ISSUE #2510: `loki.source.cloudflare`: Pulls log messages from Cloudflare using the Logpull API

### PR DESCRIPTION
**Pull request by tpaschalis**
Signed-off-by: Paschalis Tsilias <paschalis.tsilias@grafana.com>

#### PR Description
The PR adds a loki.source.cloudflare component that can read logs from Cloudflare's Logpull API and forward them to other `loki.*` components.

The testing was limited, mainly porting over what Promtail did by making use of a mock Cloudflare client, but I hope it gives us enough confidence to move ahead with the component

#### Which issue(s) this PR fixes
Fixes #2510 

#### Notes to the Reviewer
Nothing in particular.

#### PR Checklist

- [X] CHANGELOG updated
- [X] Documentation added
- [X] Tests updated

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -26,6 +26,7 @@ require (
 	github.com/google/go-jsonnet v0.18.0
 	github.com/google/uuid v1.3.0
 	github.com/gorilla/mux v1.8.0
+	github.com/grafana/cloudflare-go v0.0.0-20230110200409-c627cf6792f2
```
Comment (tpaschalis) [inline go.mod:29]: Recently, [this](https://github.com/grafana/loki/commit/5eba0286563c9de6df1273e68b30bf5b58e54ba3) commit from @mar4uk removed the Loki replace directives for cloudflare.

Next time we upgrade our Loki, we'll be able to remove them as well, but for now I'm preemptively adding the fork to our go.mod to keep the same import paths as the ported code.

File: component/loki/source/cloudflare/cloudflare.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+		Ready:   c.target.Ready(),
+		Labels:  lbls,
+		Details: c.target.Details(),
+	}
+}
+
+type targetDebugInfo struct {
+	Ready   bool              `river:"ready,attr"`
+	Labels  map[string]string `river:"labels,attr"`
+	Details map[string]string `river:"target_info,attr"`
```
Comment (tpaschalis) [inline component/loki/source/cloudflare/cloudflare.go:143]: We can have each field on its own here, I was just lazy at this point. Will see how it looks on the UI view.

File: component/loki/source/cloudflare/types.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+// Config defines how to create a Cloudflare logs target.
+type Config struct {
+	APIToken   string            `river:"api_token,attr"`
+	ZoneID     string            `river:"zone_id,attr"`
+	Labels     map[string]string `river:"labels,attr,optional"`
+	Workers    int               `river:"workers,attr,optional"`
+	PullRange  time.Duration     `river:"pull_range,attr,optional"`
+	FieldsType string            `river:"fields_type,attr,optional"`
```
Comment (rfratto) [inline component/loki/source/cloudflare/types.go:19]: It feels a bit weird for the bulk of the config to be in a `config` block. Was there a reason to do this vs keeping a flat list of attributes in the arguments?

File: component/loki/source/cloudflare/cloudflare.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.handler:
+			c.mut.RLock()
+			for _, receiver := range c.fanout {
+				receiver <- entry
+			}
+			c.mut.RUnlock()
+		}
+	}
```
Comment (rfratto) [inline component/loki/source/cloudflare/cloudflare.go:98]: We've done this pattern so many times now, I wonder if it's time for an abstraction to make it a little easier 🤔

File: component/loki/source/cloudflare/cloudflare.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		level.Info(c.opts.Logger).Log("msg", "loki.source.cloudflare component shutting down, stopping the target")
+		c.target.Stop()
```
Comment (rfratto) [inline component/loki/source/cloudflare/cloudflare.go:84]: There's a race condition here with `Update`; you need to grab the mutex before stopping the target.

File: component/loki/source/cloudflare/cloudflare.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+	return targetDebugInfo{
+		Ready:   c.target.Ready(),
+		Labels:  lbls,
+		Details: c.target.Details(),
+	}
+}
+
+type targetDebugInfo struct {
+	Ready   bool              `river:"ready,attr"`
+	Labels  map[string]string `river:"labels,attr"`
```
Comment (rfratto) [inline component/loki/source/cloudflare/cloudflare.go:142]: Are the labels here any different than what is configured in the arguments? If they're always going to be exactly the same, I'm not sure if there's much value in putting them here.

File: component/loki/source/cloudflare/cloudflare.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
+package cloudflare
```
Comment (tpaschalis) [inline component/loki/source/cloudflare/cloudflare.go:1]: TODO: Check if there's a good way to test the component itself without making it too messy with the mocks.

**Review by karengermond — APPROVED** [review]
LGTM!

File: component/loki/source/cloudflare/types.go (hunk 3 above)
Comment (tpaschalis) [reply component/loki/source/cloudflare/types.go:19]: No, let's go with the flatter Arguments structure.

File: component/loki/source/cloudflare/cloudflare.go (hunk 4 above)
Comment (tpaschalis) [reply component/loki/source/cloudflare/cloudflare.go:151]: I've got something in mind, I'd prefer to have it as a separate PR.

**Pull request merged**